{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2329d6b1-393d-4bf5-b582-0bc3779257c6",
   "metadata": {},
   "source": [
    "# Download\n",
    "\n",
    "The raw datasets are available for download from \n",
    "\n",
    "https://figshare.com/projects/Tabula_Muris_Transcriptomic_characterization_of_20_organs_and_tissues_from_Mus_musculus_at_single_cell_resolution/27733 \n",
    "\n",
    "10X Single-cell RNA-seq data from microfluidic emulsion (v2)\n",
    "\n",
    "SS2 Single-cell RNA-seq data from Smart-seq2 sequencing of FACS sorted cells (v2)\n",
    "\n",
    "# Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "87fcbdc2-9777-47e6-a43a-a423dbb1a127",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "be10b9b0-84b9-476c-8ca0-45937246997a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-2-d7e6a7c8a664>:33: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead\n",
      "  genes = adata_SS2_all.var.index & adata_SS2.var.index\n",
      "<ipython-input-2-d7e6a7c8a664>:71: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead\n",
      "  genes = adata_10X_all.var.index & adata_10X.var.index\n",
      "... storing 'tissue' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'batch' as categorical\n",
      "... storing 'tissue' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'batch' as categorical\n",
      "... storing 'tissue' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'batch' as categorical\n"
     ]
    }
   ],
   "source": [
    "data_path = '../datasets/preprocessed/'\n",
    "\n",
    "if not os.path.exists(data_path):\n",
    "    \n",
    "    os.makedirs(data_path)\n",
    "\n",
    "path_SS2 = '../datasets/raw/Tabula_Muris/FACS/'\n",
    "path_10X = '../datasets/raw/Tabula_Muris/droplet/'\n",
    "\n",
    "meta_SS2 = pd.read_csv('../datasets/raw/Tabula_Muris/annotations_facs.csv', \n",
    "                       keep_default_na = False)\n",
    "meta_10X = pd.read_csv('../datasets/raw/Tabula_Muris/annotations_droplet.csv', \n",
    "                       keep_default_na = False)\n",
    "\n",
    "# SS2\n",
    "for tissue in ['Aorta', 'Bladder', 'Brain_Myeloid', 'Brain_Non-Myeloid', 'Diaphragm', 'Fat', 'Heart',  \n",
    "               'Kidney', 'Large_Intestine', 'Limb_Muscle', 'Liver', 'Lung', 'Mammary_Gland', 'Marrow', \n",
    "               'Pancreas', 'Skin', 'Spleen', 'Thymus', 'Tongue', 'Trachea']:\n",
    "\n",
    "    # Read SS2 cell-by-gene counts\n",
    "    adata_SS2 = sc.read_csv(os.path.join(path_SS2, \"%s-counts.csv\" % tissue)).transpose()\n",
    "    ERCC_idx = pd.Series(adata_SS2.var.index).str.startswith('ERCC')\n",
    "    cell_idx = adata_SS2.obs.index.isin(meta_SS2[(meta_SS2.cell_ontology_class != 0) & \n",
    "                                                         (meta_SS2.cell_ontology_class != '')].cell)\n",
    "    adata_SS2 = adata_SS2[cell_idx, -ERCC_idx]\n",
    "    \n",
    "    if tissue == 'Aorta':\n",
    "        \n",
    "        adata_SS2_all = adata_SS2.copy()\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        genes = adata_SS2_all.var.index & adata_SS2.var.index\n",
    "        adata_SS2_all = adata_SS2_all[:, genes].concatenate(adata_SS2[:, genes], index_unique = None)\n",
    "\n",
    "# 10X\n",
    "for tissue in ['Bladder', 'Heart_and_Aorta', 'Kidney', 'Limb_Muscle', 'Liver', 'Lung', 'Mammary_Gland', \n",
    "               'Marrow', 'Spleen', 'Thymus', 'Tongue', 'Trachea']:\n",
    "\n",
    "    # Read 10X cell-by-gene counts\n",
    "    channels = sorted(set(meta_10X[meta_10X.tissue == tissue].channel))\n",
    "    \n",
    "    for i, channel in enumerate(channels):\n",
    "        \n",
    "        if i == 0:\n",
    "            \n",
    "            adata_10X = sc.read_10x_mtx(path_10X + '/%s-%s/' % (tissue, channel), \n",
    "                                        var_names = 'gene_symbols', cache = False)\n",
    "            adata_10X.obs.index = channel + \"_\" + adata_10X.obs.index\n",
    "            adata_10X.obs.index = adata_10X.obs.index.map(lambda x: x[:-2])\n",
    "            cell_idx = adata_10X.obs.index.isin(meta_10X[(meta_10X.cell_ontology_class != 0) &\n",
    "                                                         (meta_10X.cell_ontology_class != '')].cell)\n",
    "            adata_10X = adata_10X[cell_idx, :]\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            tmp = sc.read_10x_mtx(path_10X + '/%s-%s/' % (tissue, channel), \n",
    "                                  var_names = 'gene_symbols', cache = False)\n",
    "            tmp.obs.index = channel + \"_\" + tmp.obs.index\n",
    "            tmp.obs.index = tmp.obs.index.map(lambda x: x[:-2])\n",
    "            cell_idx = tmp.obs.index.isin(meta_10X[(meta_10X.cell_ontology_class != 0) &\n",
    "                                                   (meta_10X.cell_ontology_class != '')].cell)\n",
    "            adata_10X = adata_10X.concatenate(tmp[cell_idx, :], index_unique=None)\n",
    "            \n",
    "    if tissue == \"Bladder\":\n",
    "        \n",
    "        adata_10X_all = adata_10X.copy()\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        genes = adata_10X_all.var.index & adata_10X.var.index\n",
    "        adata_10X_all = adata_10X_all[:, genes].concatenate(adata_10X[:, genes], index_unique = None)\n",
    "\n",
    "meta_SS2 = meta_SS2[meta_SS2.cell.isin(adata_SS2_all.obs.index)][['cell', 'tissue', 'cell_ontology_class']].set_index('cell')\n",
    "meta_SS2['batch'] = 'SS2'\n",
    "meta_SS2 = meta_SS2.rename(columns = {'cell_ontology_class': 'celltype'})\n",
    "adata_SS2_all.obs = meta_SS2.loc[adata_SS2_all.obs.index]\n",
    "\n",
    "meta_10X = meta_10X[meta_10X.cell.isin(adata_10X_all.obs.index)][['cell', 'tissue', 'cell_ontology_class']].set_index('cell')\n",
    "meta_10X['batch'] = '10X'\n",
    "meta_10X = meta_10X.rename(columns = {'cell_ontology_class': 'celltype'})\n",
    "adata_10X_all.obs = meta_10X.loc[adata_10X_all.obs.index]\n",
    "\n",
    "adata_SS2_all.write(filename = os.path.join(data_path, 'TM_SS2.h5ad'))\n",
    "adata_10X_all.write(filename = os.path.join(data_path, 'TM_10X.h5ad'))\n",
    "\n",
    "sc.concat((adata_SS2_all, adata_10X_all)).write(filename = os.path.join(data_path, 'TM_full.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1624cf6c-19ae-4ed4-922b-7d60ff1d96eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the following code in R to convert h5ad to h5seurat which Seurat handles\n",
    "# Check current directory first\n",
    "# library(SeuratDisk)\n",
    "# Convert('TM_full.h5ad', 'h5seurat', overwrite = TRUE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
